state-chain/chains/src/sol/instruction_builder.rs

```diff
@@ -340,13 +343,38 @@ impl SolanaInstructionBuilder {
 			sys_var_instructions(),
 		).with_remaining_accounts(ccm_accounts.remaining_account_metas())];
 
-		// TODO: Complete in PRO-1479
 		Self::finalize(
 			instructions,
 			nonce_account.into(),
 			agg_key.into(),
 			compute_price,
-			COMPUTE_LIMIT,
+			Self::calculate_gas_limit(gas_budget, compute_price, SolAsset::SolUsdc),
+		)
+	}
+
+	fn calculate_gas_limit(
+		gas_budget: SolAmount,
+		compute_price: SolAmount,
+		asset: SolAsset,
+	) -> SolComputeLimit {
+		let budget_after_signature = gas_budget.saturating_sub(LAMPORTS_PER_SIGNATURE);
+		if compute_price == 0 {
+			return DEFAULT_COMPUTE_UNITS_PER_CCM_TRANSFER;
+		}
+		let compute_budget =
+			// Budget is in lamports, compute price is in CU/microlamport
+			sp_std::cmp::min(
+				MAX_COMPUTE_UNITS_PER_TRANSACTION as u128,
+				(budget_after_signature as u128 * MICROLAMPORTS_PER_LAMPORT as u128)
+					/ (compute_price as u128),
```

Comment:
If compute_price really is "compute units per microlamport", then this should be a multiplication, not a divison

Reply:
The comment is what's wrong, `compute_price` is microlamports per compute unit. I'll fix it.